gtars/src/ailist/core.rs

```diff
@@ -0,0 +1,182 @@
+use std::fmt;
+
+pub struct Interval {
+    pub start: u32,
+    pub end: u32,
+}
+
+impl fmt::Display for Interval {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "({}, {})", self.start, self.end)
+    }
+}
+
+///
+/// The Augmented Interval List (AIList), enumerates intersections between a query interval q and an interval set R.
+///
+pub struct AIList {
+    starts: Vec<u32>,
+    ends: Vec<u32>,
+    max_ends: Vec<u32>,
+    header_list: Vec<usize>,
+}
+
+impl AIList {
+    ///
+    /// Create a new AIList struct
+    ///
+    /// # Arguments
+    /// - intervals: list of intervals to create from
+    ///
+    /// # Returns
+    /// - AIList struct
+    pub fn new(intervals: &mut Vec<Interval>, minimum_coverage_length: usize) -> AIList {
+        // in the future, clone and sort...
+        intervals.sort_by_key(|key| key.start);
+
+        let mut starts: Vec<u32> = Vec::new();
```

Comment:
```suggestion
        let mut starts = Vec::with_capacity(intervals.len());
```

Same for all the other inits. You know the size so this saves space and allocations. 

Additionally, avoid specifying types where possible. Typically you can just have types in fn signatures and that's it. It makes refactoring a lot easier when the types are inferred.

Reply:
Are there ever downsides to using `::with_capacity();`?

Reply:
Not really, it should be pretty much all upside!

1. one allocation for the whole thing
2. No "extra" space that is the result of growing the space in powers of two when you use push.

The only downside would be reserving too much capacity and not using it, which shouldn't be the case here.